File: languages/go/fannkuch.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func fannkuch(n int) (int, int) {
	perm1 := make([]int, n)
	perm := make([]int, n)
	count := make([]int, n)
	for i := range perm1 {
		perm1[i] = i
	}
	maxFlips := 0
	checksum := 0
	permIdx := 0
	r := n

	for {
		for r != 1 {
			count[r-1] = r
			r--
		}

		copy(perm, perm1)
		flips := 0
		for k := perm[0]; k != 0; k = perm[0] {
			for i, j := 0, k; i < j; i, j = i+1, j-1 {
				perm[i], perm[j] = perm[j], perm[i]
			}
			flips++
		}

		if flips > maxFlips {
			maxFlips = flips
		}
		if permIdx%2 == 0 {
			checksum += flips
		} else {
			checksum -= flips
		}

		// Generate the next permutation.
		for {
			if r == n {
				return maxFlips, checksum
			}
			first := perm1[0]
			copy(perm1[:r], perm1[1:r+1])
			perm1[r] = first
			count[r]--
			if count[r] > 0 {
				break
			}
			r++
		}
		permIdx++
	}
}

func main() {
	n := 7
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil {
			n = v
		}
	}
	maxFlips, checksum := fannkuch(n)
	fmt.Println(checksum)
	fmt.Printf("Pfannkuchen(%d) = %d\n", n, maxFlips)
}

File: languages/kotlin/fannkuch.kt
fun fannkuch(n: Int): Pair<Int, Int> {
    val perm1 = IntArray(n) { it }
    val perm = IntArray(n)
    val count = IntArray(n)
    var maxFlips = 0
    var checksum = 0
    var permIdx = 0L  // counts up to n!-1 → needs 64-bit for n >= 13
    var r = n

    while (true) {
        while (r != 1) {
            count[r - 1] = r
            r--
        }

        System.arraycopy(perm1, 0, perm, 0, n)
        var flips = 0
        var k = perm[0]
        while (k != 0) {
            var i = 0
            var j = k
            while (i < j) {
                val t = perm[i]; perm[i] = perm[j]; perm[j] = t
                i++; j--
            }
            flips++
            k = perm[0]
        }

        if (flips > maxFlips) maxFlips = flips
        checksum += if (permIdx % 2 == 0L) flips else -flips

        // Generate the next permutation.
        while (true) {
            if (r == n) return Pair(maxFlips, checksum)
            val first = perm1[0]
            for (i in 0 until r) perm1[i] = perm1[i + 1]
            perm1[r] = first
            count[r]--
            if (count[r] > 0) break
            r++
        }
        permIdx++
    }
}

fun main(args: Array<String>) {
    val n = if (args.isNotEmpty()) args[0].toInt() else 7
    val (maxFlips, checksum) = fannkuch(n)
    println(checksum)
    println("Pfannkuchen($n) = $maxFlips")
}
